### ISSUE TITLE
Issue propagating gradients through tf.while_loop

### ISSUE THREAD

**Issue by fjanoos** [tensorflow/tensorflow#16504]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu wheezy

- **TensorFlow installed from (source or binary)**:
pip

- **TensorFlow version (use command below)**:
tf.VERSION = 1.4.0
tf.GIT_VERSION = v1.4.0-4-g9283868
tf.COMPILER_VERSION = v1.4.0-4-g9283868
Sanity check: array([1], dtype=int32)

- **Python version**: 
3.5

### Describe the problem

I've found a few issues when trying to propagate gradients through tf.while_loops.

One issue is ops like this inside the loop body break gradients 
```k = tf.Print( k + 1, [k + 1, eta, loss( w_n ), chg_w, G_inf], 'EG:: k, eta, loss(w), chg_w, G_inf = ' )```

but more concerningly, conjoined conditions such as this:
```tf.logical_and( k < max_its-1, chg_w > tol  )```
or even this
``` tf.cast( max_its-k, DTYPE) *(chg_w - tol)```
breaks the differentiablity across the while loop.

### Source code / logs
``` python
tf.reset_default_graph()
sess = tf.InteractiveSession()
g = tf.Graph().as_default()

max_its = 10
tol = 1e-3

c = tf.constant( np.arange(100), dtype=DTYPE)
w = tf.Variable( initial_value=np.ones(100),  dtype=DTYPE)/100
k = tf.Variable( 0, dtype=tf.int32 )
chg_w = tf.constant( np.inf, dtype=DTYPE )


def _eg_step( k, w, chg_w): 
    grad = tf.gradients( -tf.reduce_sum( w * c ) , w )[0]
    w_n = w * tf.exp( -0.1  * grad )
    w_n = w_n / tf.reduce_sum( w_n ) 
    chg_w = tf.reduce_sum( tf.abs( w_n - w) ) / tf.reduce_sum( tf.abs( w ) )
    k = k + 1
    **# !! this busts the differentiablity !!
    # k = tf.Print( k + 1, [k + 1, eta, loss( w_n ), chg_w, G_inf], 'EG:: k, eta, loss(w), chg_w, G_inf = ' )**
    return k, w_n, chg_w

def _continue_cond( k, w, chg_w, *args ):
    **# NOTE either of this conjoined conditions
    #        tf.logical_and( k < max_its-1, chg_w > tol  )
    # OR     tf.cast( max_its-k, DTYPE) *(chg_w - tol)
    # do no propagate gradients correctly**
    return  k < max_its # tf.logical_and( k < max_its-1, chg_w > tol  )

k, w, chg_w = tf.while_loop(
    cond=_continue_cond,  body=_eg_step,
    loop_vars=[k, w, chg_w],
    name='while_loop', parallel_iterations=1
)

# see if the gradient is propagated
tf.gradients( w, c)

```

**Comment by drpngx** [comment]
I think I have observed the problem with `Print`. Thank you for having that simple repro case! Do you think you could submit a PR to fix this?

Regarding the logical conditions, could you be more specific about what gradients gets zero?

CC @zffchen78 @skye

**Comment by fjanoos** [comment]
@drpngx  what is a PR ? 

regarding the logical conditions:

If the _continue_cond is a conjoined condition , the gradient of output[round_t]/<something> becomes [None], where as if the condition is  a simple expression grad output[round_t]/<something> is correctly evaluated.

For example - in the above example if
```
def _continue_cond( k, w, chg_w, *args ):
    return  k < max_its
```
then tensorflow can correctly evaluate `tf.gradients( w, c)`

But if I make it:
```
def _continue_cond( k, w, chg_w, *args ):
    return tf.logical_and( k < max_its-1, chg_w > tol  )
    # OR     tf.cast( max_its-k, DTYPE) *(chg_w - tol)
```

then `tf.gradients( w, c) = [None]`

**Comment by drpngx** [comment]
Sorry, I meant a pull request, ie a bugfix.

**Comment by fjanoos** [comment]
I'm not sure I have a bug fix - since I don't know what the root of the problem is  :)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ccolby** [comment]
The candidate fix is scheduled for a design review in a couple of weeks. It's going through design review because it changes the behavior of tf.gradients.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ccolby** [comment]
The change passed design review and will go in soon. Someone can reassign this issue to me. (I cannot do it.)

**Comment by skye** [comment]
Unfortunately we can currently only assign to TF org members :\

**Comment by ccolby** [comment]
This issue should be fixed in the master branch.
Search for testIssue16504 in [control_flow_ops_py_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/control_flow_ops_py_test.py).

**Comment by drpngx** [comment]
Woohoo!
